src/Compilers/CSharp/Portable/Symbols/AnonymousTypes/AnonymousTypeManager.Templates.cs

```diff
@@ -163,60 +156,187 @@ private void CheckSourceLocationSeen(AnonymousTypePublicSymbol anonymous)
+                }
+
+                // Adjust template location if the template is owned by this manager
+                if (ReferenceEquals(template.Manager, this))
+                {
+                    template.AdjustLocation(typeDescr.Location);
                 }
+
+                Debug.Assert(typeParameters.Length == template.TypeParameters.Length);
+                return typeParameters.Length == 0 ?
+                    template :
+                    template.Construct(typeParameters);
             }
-            var typeArgumentsBuilder = ArrayBuilder<TypeWithAnnotations>.GetInstance(nTypeArguments);
-            for (int i = 0; i < nTypeArguments; i++)
+
+            static bool allValidTypeArguments(AnonymousTypeDescriptor typeDescr)
             {
-                typeArgumentsBuilder.Add(fields[i].TypeWithAnnotations);
+                var fields = typeDescr.Fields;
+                int n = fields.Length;
+                for (int i = 0; i < n - 1; i++)
+                {
+                    if (!isValidTypeArgument(fields[i]))
+                    {
+                        return true;
+                    }
+                }
+                var returnParameter = fields[n - 1];
+                return !returnParameter.Type.IsVoidType() && !isValidTypeArgument(returnParameter);
             }
-            var typeArguments = typeArgumentsBuilder.ToImmutableAndFree();
-            var template = SynthesizeDelegate(parameterCount: fields.Length - 1, refKinds, returnsVoid, generation);
 
-            Debug.Assert(typeArguments.Length == template.TypeParameters.Length);
-            return typeArguments.Length == 0 ?
-                template :
-                template.Construct(typeArguments);
+            static bool isValidTypeArgument(AnonymousTypeField field)
+            {
+                return field.Type is { } type &&
+                    !type.IsPointerOrFunctionPointer() &&
+                    !type.IsRestrictedType();
+            }
+
+            static SynthesizedDelegateKey getTemplateKey(AnonymousTypeDescriptor typeDescr, ImmutableArray<TypeParameterSymbol> typeParameters)
+            {
+                if (typeParameters.Length > 0)
+                {
+                    var typeMap = new TypeMap(typeParameters, IndexedTypeParameterSymbol.Take(typeParameters.Length), allowAlpha: true);
+                    typeDescr = typeDescr.SubstituteTypes(typeMap, out bool changed);
+                    Debug.Assert(changed);
+                }
+                return new SynthesizedDelegateKey(typeDescr);
+            }
+        }
+
+        private static ImmutableArray<TypeParameterSymbol> GetReferencedTypeParameters(AnonymousTypeDescriptor typeDescr)
+        {
+            var referenced = PooledHashSet<TypeParameterSymbol>.GetInstance();
+            foreach (var field in typeDescr.Fields)
+            {
+                field.Type.VisitType(
```

Comment:
>field.Type.VisitType [](http://example.com/codeflow?start=16&length=20)

I think we should call ```VisitType``` on ```field.TypeWithAnnotations``` and dig into custom modifiers as well. ```VistType``` doesn't handle custom modifiers at the moment, but that is getting fixed in [https://github.com/dotnet/roslyn/pull/58288](https://github.com/dotnet/roslyn/pull/58288). We can either copy that fix into this PR, or also supply a filter for ```TypeWithAnnotation``` that does the same thing that [https://github.com/dotnet/roslyn/pull/58288](https://github.com/dotnet/roslyn/pull/58288) does for custom modifiers. And we need to test this, there are test cases in the same PR that can be used  for inspiration.